torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+    auto end2 = buf_ranges.at(b2).second;
+    return std::get<2>(end1).at(0) < std::get<2>(end2).at(0);
+  };
+
+  // Map intermediate buffers to the most recent used memory if any.
+  std::unordered_set<BufPtr> mm;
+  std::list<BufPtr> mm_free;
+  std::unordered_map<BufPtr, BufPtr> b2m;
+  std::vector<std::pair<BufPtr, BufPtr>> b2m_ret;
+
+  std::vector<BufPtr> buf_to_release;
+  for (auto buf : bufs_sorted) {
+    auto start = buf_ranges.at(buf).first;
+    auto end = buf_ranges.at(buf).second;
+
+    // Release memory for buffers whose live range ends before the creation time
+    // of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
+    for (auto& mapped : b2m) {
+      auto buf_mapped = mapped.first;
+      auto end_buf_mapped = buf_ranges.at(buf_mapped).second;
+      if (std::get<2>(end_buf_mapped).at(0) < std::get<2>(start).at(0)) {
+        buf_to_release.push_back(buf_mapped);
+      }
+    }
+    std::sort(
+        buf_to_release.begin(),
+        buf_to_release.end(),
+        sorting_function_by_end_time);
+    for (auto& buf_rl : buf_to_release) {
+      mm_free.push_front(b2m[buf_rl]);
+      b2m.erase(buf_rl);
+    }
+
+    // If the buf has dynamic shapes, we'll skip it (i.e., allocate memory for
+    // it, and there are no future reuses on its memory).
+    // TODO: reuse memory for bufs with dynamic shapes
+    if (!bufSize(buf)) {
+      b2m_ret.emplace_back(std::make_pair(buf, buf));
+      continue;
+    }
+
+    bool allocated = false;
+    // Check whether there are free memories that this buf can reuse.
+    for (auto it = mm_free.begin(); it != mm_free.end(); it++) {
+      auto m = *it;
+      TORCH_INTERNAL_ASSERT(bufSize(buf) != 0);
+      TORCH_INTERNAL_ASSERT(bufSize(m) != 0);
+      if (bufSize(m) >= bufSize(buf)) {
+        b2m[buf] = m;
+        b2m_ret.emplace_back(std::make_pair(buf, m));
+        allocated = true;
+        mm_free.erase(it);
+        break;
+      }
+    }
+
+    // If there are no memories to reuse, we'll have to allocate new memory for
+    // it.
+    if (!allocated) {
+      mm.insert(buf);
+      b2m[buf] = buf;
+      b2m_ret.emplace_back(std::make_pair(buf, buf));
+    }
+  }
+
+  return b2m_ret;
+}
+
+StmtPtr insertAllocFree(
+    std::vector<std::pair<BufPtr, BufPtr>>& b2m,
+    StmtPtr stmt) {
   BlockPtr b = to<Block>(stmt);
   if (!b) {
     b = alloc<Block>(std::vector<StmtPtr>({stmt}));
   }
 
   // Insert allocations and frees for temporary buffers at global scope.
-  for (auto buf : interm_bufs) {
-    b->prepend_stmt(alloc<Allocate>(buf));
-    b->append_stmt(alloc<Free>(buf));
+  for (auto rit = b2m.rbegin(); rit != b2m.rend(); ++rit) {
+    if (rit->first == rit->second) {
+      BufPtr buf = rit->first;
+      b->prepend_stmt(alloc<Allocate>(buf));
+      b->append_stmt(alloc<Free>(buf));
+    } else {
+      b->prepend_stmt(alloc<PlacementAllocate>(rit->first, rit->second));
```

Comment:
We're inserting PlacementAllocate to the beginning of the block. I think it's technically correct with the current implementation, but it breaks a neat mental model that it is equivalent to `"Free(old); Allocate(new);"`. Would it be possible to insert it after the last use of the old buffer and before the first use of the new one?

Reply:
As we discussed in person, we'll move all `Allocate/PlacementAllocate/Free` from the beginning to where they should be in a separate PR.